### DISCUSSION TITLE
Going forward with the OpenAndroidInstaller

### DISCUSSION THREAD

**Discussion by tsterbak** [openandroidinstaller-dev/openandroidinstaller#638] [Announcements]
Hey everyone, sorry for the long quietness. I was tied up with other things. Now I'm back to put more work in the OpenAndroidInstaller.

I see several issues with the current approach and have a few ideas on how to further develop the installer. And I would like to hear your opinion about these as well as if you have other ideas or suggestions.

The major issues with the current approach I took with the installer are:
1. TWRP has quite some quirks and a lot of newer devices are not (yet?) supported. Development here seems slower since LineageOS released their own recovery.
2. Maintaining a proper cross-platform desktop application with usb-support is difficult and unstable.
3. Maintaining easy to use and reasonably safe installation instructions and requiremend checks.

The options I see going forward are:
1. **Focus** stronger (or completely) **on devices officially supported by LineageOS**. Then **drop TWRP** (and lose the scripting part of the automation) and rely on the LOS recovery (or similar recoveries from e/os/, etc.). This way we could offer basically the same support set as LineageOS and rely on their installation instructions (mostly, some problably will be hard to support).
2. Moving to **a webbased installer**, similar to https://github.com/kdrag0n/android-webinstall . With [webabd](https://webadb.com/) and [fastboot.js](https://kdrag0n.github.io/fastboot.js/) we could make this easier for people and also support automated downloads for ROMs and recovery.
3. Do both or combinations there off.
4. **Continue as is**. Add support for more TWRP-based recoveries (like Ofox) and try to support more devices and keep it as stable as possible.

I would love to hear your ideas and opinions to these ideas and issues. What do you think?

**Comment by tsterbak** [comment]
@freechelmi @MagicLike @anon1892

**Reply by pedrocostatolottit-art** [reply]
mano, porque eu estou contribuindo com isso?minha conta estava inativa, eu nem recebi convite

**Comment by freechelmi** [comment]
Yes I had the same questioning when I saw that eOS installer is going web based in the next step 

The nice part is that going web based allow to go fat app quickly with electron or similar

**Comment by kevATin** [comment]
I like the idea of the installer being web-based, but only if it works consistently on Linux, Windows, macOS, and maybe also Android.
And there should be ways to set up the tool without relying on the official web server, either by making it easy to self-host the website or by packaging it with electron or something.

But I feel like there are so many projects out there that duplicate efforts like this.

There is non-Android postmarketOS https://postmarketos.org/ which supports manual installation via fastboot, and the UBports installer, and they also have their own web installer https://flash.postmarketos.org/ which is of course open source https://gitlab.com/postmarketOS/webflash 

Then there is non-Android UBports https://ubports.com/en/ which also supports manual installation but recommends their UBports installer https://devices.ubuntu-touch.io/installer/ which does not appear to be web based and is open source too https://github.com/ubports/ubports-installer 

There is also non-Android AsteroidOS https://asteroidos.org/ for smartwatches which only supports manual adb/fastboot installation.

And eOS has their own installer too https://doc.e.foundation/easy-installer which is open source.

GrapheneOS appears to have a web installer https://grapheneos.org/install/web 

Even Google itself does https://flash.android.com/

And there are probably dozens of smaller projects that try to make their own installers.

Surely there must be significant overlap between all those projects so that it would make sense to share installer-tool efforts.

Personally I'm a huge fan of Calamares https://calamares.io/ / https://github.com/calamares/calamares because if there is one common installer standard for Linux distros, then that frees up resources for other efforts and we really don't need everyone to reinvent the installer.

Maybe it would make sense to create something similar to that but for Android-based devices. 
The official tool could provide access to the most common OS images and features, and other projects could maintain their own fork of the tool, disable the features they don't want, remove some OS images, add some OS images, put on some branding and host it on their own website as their official installer.
That way if someone adds some useful feature they can upstream it and it'll benefit all projects using this tool, instead of everyone doing their own thing.

**Reply by freechelmi** [reply]
>only if it works consistently on Linux, Windows, macOS, and maybe also Android. 

it needs webusb support which mozilla thinks is too dangerous too support and apple , well it's apple :-) 

> There is non-Android postmarketOS https://postmarketos.org/ 
[...]
> Even Google itself does https://flash.android.com/

I feel the same , we should be able to implement some "libflashdevice.js" which should be reusable by all those projects

**Reply by freechelmi** [reply]
We can add the eOS web installer to the list , ( [list of supported devices ](https://doc.e.foundation/easy-installer)) also which can be a big win for newcomers , some manual step as the LOS recovery cannot be automated

https://e.foundation/installer/

**Reply by tsterbak** [reply]
Yeah, the new e/OS/ webinstaller looks awesome! I was thinking of something similar. My main issue with e/OS/ is, that the number of supported devices, especially in the installer is so small. But it works pretty well if it does :)

**Reply by freechelmi** [reply]
git project of the eos-installer https://gitlab.e.foundation/e/devices/eos-installer

**Comment by Klaasstap** [comment]
Apart for some people, I think flashing one's phone to have another operating system is something you don't want to do regularly.
So, TWRP is not really necessary, I believe. Someone who chooses /e/OS will be happy with it's own recovery and someone who wants LineageOS will be happy using it's recovery. When people do want to switch, they can rely to all the guides that are out there.
Would this approach fit in the development of a web based installer?

By the way: it's a pity only a few people know about flashing.

**Reply by tsterbak** [reply]
TWRP is not flashed (in most cases, some older samsung phones require it...) but only booted. So ROMs like LineageOS and eOS (I suppose) will replace the recovery with their recovery at first boot. OTA updates with the recoveries of the respective ROMs work then.
TWRP allows to script all the necessary steps you need to take in recovery mode. Thats why we require it.
It would also be possible to have a lineageOS recovery based version of the installer (that would also work webbased, I guess eOS is also working on this). But then it is merely a tool to tell you what to do and boot the recovery... Not much easier than just follow the instructions on the website, I guess.

> By the way: it's a pity only a few people know about flashing.

I agree! :)

**Comment by ranaji10** [comment]
@tsterbak It's been a while since this discussion was active, and since I'm currently building something along similar lines, I thought I'll use this thread to do a temperature check. 

Can the idea of eos-installer be extended further? Currently it solves for "how do I safely flash eOS with browser automation?" 

I'm currently building a tool to solve the problem of "given any device and any recipe, can I verify it's safe to flash?" 

Would a general verdict like that be useful? Is that something that OpenAndroidInstaller is exploring and if not, what would the tool have to get right, before it could become useful to OAI?
